src/core/io/writer/h5md/h5md_core.cpp

```diff
@@ -0,0 +1,707 @@
+/*
+  Copyright (C) 2010,2011,2012,2013,2014,2015,2016 The ESPResSo project
+  Copyright (C) 2002,2003,2004,2005,2006,2007,2008,2009,2010 
+    Max-Planck-Institute for Polymer Research, Theory Group
+  
+  This file is part of ESPResSo.
+  
+  ESPResSo is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+  
+  ESPResSo is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+  
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>. 
+*/
+
+
+#include "h5md_core.hpp"
+
+
+namespace writer {
+namespace h5md {
+
+
+/* Constructor for the File class. */
+File::File(std::string const &filename, std::string const &script_name)
+{
+    boost::filesystem::path script_path(script_name);
+    this->absolute_script_path = boost::filesystem::canonical(script_path);
+    /* Store the filename in a member variable. */
+    this->user_filename = filename;
+    /* Get number of local particles. */
+    this->n_local_part = cells_get_n_particles();
+    if(!(this->n_local_part > 0)) {
+        throw std::runtime_error("Please first set up particles before initializing the H5md object.");
+    } 
+    /* Check if a file with given filename exists. */
+    bool file_exists = this->check_file_exists(filename);
+    /* If it exists, check for the H5MD structure. */
+    if (file_exists)
+    {
+        this->has_H5MD_structure = check_for_H5MD_structure(filename);
+    }
+    if (file_exists && this->has_H5MD_structure)
+    {
+        /*
+         * If the file exists and has a valid H5MD structure, lets create a new file and copy the hdf5 objects.
+         * This has the advantage, that the new file can just be deleted if the simulation crashes at some point and we
+         * still have a valid trajectory, we can start from.
+        */
+        std::string temp_name = filename + "_tmp";
+        this->lastfile = h5xx::file(filename, MPI_COMM_WORLD, MPI_INFO_NULL,
+                                     h5xx::file::out);
+        this->h5md_file = h5xx::file(temp_name, MPI_COMM_WORLD, MPI_INFO_NULL,
+                                     h5xx::file::out);
+        /* particles -- atoms -- box -- edges */
+        H5Ocopy(lastfile.hid(), "/particles", this->h5md_file.hid(), "/particles", H5P_DEFAULT,
```

Comment:
`nano-dimer` is implemented in Lua and OpenCL and is not easy to navigate at first. I tried to share my first steps in it in a blog post if you are interested: http://pdebuyl.be/blog/2015/opencl-nanomotors.html

Specifically for snapshotting:
1. The file `nanomotor/snapshot.lua` contains the generic mechanism for snapshotting. It relies on objects having defined a `snapshot` and `restore` function.
2. For specific objects, such as trajectories and observables, the functions are defined and explicity save the object's state to disk (see e.g. `nanomotor/thermo.lua`) that defines the storage of the thermodynamic observables.
3. The global state of the program collects all this information. This is implemented in `nanomotor/snapshot.lua` that takes care of a "domain" and is provided with observables (last function in the file).

The snapshot filename and interval are defined in the configuration file of a simulation. The snapshot file is different from the trajectory file in the sense that, for instance, the solvent information is stored fully for every snapshot. If I understand well, the snapshots are also deviating from the HDF5 definition.

A snapshot is only deleted after the next snapshot has been fully written. This is achieved by a system call to rename `snapshot_filename.h5.tmp` to `snapshot_filename.h5` after closing the HDF5 file. (again, end of `nanomotor/snapshot.lua`).

The code can be found at https://git.colberg.org/nano-dimer/tree/